### PR TITLE
build(snap): Enable secure Kong Admin APIs in snap build

### PR DESCRIPTION
**Pull request by jim-wang-intel**
- Add installation of Kong's Admin plugins template yaml to enable Admin JWT, ACL plugins etc.
- Update the Kong daemon startup shell script to import the plugins into database
- Update the snapcraft yml to integrate the secure Kong's Admin API features

Closes: #3403

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?

kong's admin APIs at 8001 are wide open without authentication and ACLs in snap

## What is the new behavior?
Kong's admin APIs are now protected with required JWT authentication with ACLs in snap

## Does this PR introduce a breaking change?

- [ ] Yes
- [x ] No

## New Imports

- [ ] Yes
- [x ] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?

In Draft PR mode until other dependent PRs for this snap to be working:
- Feature of [securing Kong's Admin APIs from this PR](https://github.com/edgexfoundry/edgex-go/pull/3328)  MERGED
- Kong's binary upgrade to new version 2.3.x PR (https://github.com/edgexfoundry/edgex-go/pull/3415)  MERGED

Now this PR is ready.

## Other information

### CODE REVIEW

File: snap/hooks/install
Comment (bnevis-i) [inline snap/hooks/install:44]: How to update this file if it is being conditionally replaced?

File: snap/snapcraft.yaml
Comment (bnevis-i) [inline snap/snapcraft.yaml:127]: Can you listen on `127.0.0.1:8100` instead?  Snaps share the host's network stack and this exposes the port to the world.

File: snap/snapcraft.yaml
Comment (bnevis-i) [inline snap/snapcraft.yaml:656]: Comment doesn't fully explain what is going on.

File: snap/snapcraft.yaml
Comment (bnevis-i) [inline snap/snapcraft.yaml:698]: Just to point out, arm64 hasn't been fixed.

**Review by bnevis-i — CHANGES_REQUESTED** [review]

File: snap/hooks/install
Comment (jim-wang-intel) [reply snap/hooks/install:44]: not sure what is conditionally replaced? I don't think we have that capability yet.

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:127]: i think most of cases would, not sure about the corner case of "remote" configuration for consul which may need to listen not just `localhost`, correct me if I am wrong.

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:656]: not sure what you want me to add here: it is just try to copy that template yml to the proper location.

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:698]: yea, my local machine is amd64 so I won;'t be able to touch it and test it here- will leave that to @siggiskulason's PR  on upgrading to Kong's version to fix that.

File: snap/hooks/install
Comment (jim-wang-intel) [reply snap/hooks/install:44]: please describe and elaborate the use case you are thinking for "conditionally replaced"

File: snap/snapcraft.yaml
Comment (bnevis-i) [reply snap/snapcraft.yaml:656]: It doesn't say WHY secretstore-setup needs this template.

File: snap/hooks/install
Comment (bnevis-i) [reply snap/hooks/install:44]: The `cp` command is skipped if `$SNAP_DATA/config/security-secretstore-setup/res/kong-admin-config.template.yml` already exists.  If it is a template, why not overwrite it every time?

File: snap/snapcraft.yaml
Comment (bnevis-i) [reply snap/snapcraft.yaml:698]: Unfortunately, you won't be able to pass the pre-commit hook as it is.

File: snap/hooks/install
Comment (jim-wang-intel) [reply snap/hooks/install:44]: oh, that, i just refer to the same logic above when dealing withe the toml configuration file, too. see line 14 above. just to keep it consistent with that.

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:656]: got you, I'll add that, thanks.

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:698]: right- i'll rebase his PR once that is there.

File: snap/snapcraft.yaml
Comment (lenny-intel) [reply snap/snapcraft.yaml:127]: Kong is our gateway, so it needs to be exposed to the world

**Review by lenny-intel — COMMENTED** [review]
LGTM, but will leave to @tonyespy / @siggiskulason for final approval.

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml]: Rebased

**Comment by jim-wang-intel** [comment]
rebased to pull in the latest changes about the kong's binary image downloading

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:127]: yea, that is what I think it should be as well

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:659]: added

**Comment by tonyespy** [comment]
@jim-wang-intel just a quick note that we started seeing failures in our automated github-action based tests with the snap install failing because of security-proxy-setup failing.

I just built your PR and as expected, it resolves this problem.

I'll review the actual code changes later today, but wanted to let you know that I've started looking at. Thanks for your changes!

**Comment by jim-wang-intel** [comment]
> @jim-wang-intel just a quick note that we started seeing failures in our automated github-action based tests with the snap install failing because of security-proxy-setup failing.
> 
> I just built your PR and as expected, it resolves this problem.
> 
> I'll review the actual code changes later today, but wanted to let you know that I've started looking at. Thanks for your changes!

@tonyespy:  Sure, thanks for the notes!

File: snap/snapcraft.yaml
Comment (tonyespy) [reply snap/snapcraft.yaml:127]: I'm going to ask the same question as @bnevis-i, why does this port (which seems to be a read-only status API) need to be accessible from a different host? Are we actively using this status API in the new security model for the admin API?

**Review by tonyespy — COMMENTED** [review]
@jim-wang-intel for the most part this PR looks good, although there are a few comments/questions inline that I'd like to see answers for...

Also I built and installed a snap from this PR and it works as expected, however I did notice one error logged by security-proxy-setup after a reboot that caused the command to fail:

```
Starting Service for snap application edgexfoundry.security-proxy-setup...
INFO app=edgex-security-proxy-setup source=config.go:359 msg="Loaded service configuration from /var/sn\
ap/edgexfound
INFO  app=edgex-security-proxy-setup source=variables.go:357 msg="Variables override of 'SecretStore.To\
kenFile' by e
INFO app=edgex-security-proxy-setup source=variables.go:357 msg="Variables override of 'KongAuth.JWTFi\
le' by enviro
INFO app=edgex-security-proxy-setup source=variables.go:357 msg="Variables override of 'AccessTokenFil\
e' by environ
INFO app=edgex-security-proxy-setup source=secret.go:49 msg="Creating SecretClient"
INFO app=edgex-security-proxy-setup source=secret.go:56 msg="Reading secret store configuration and authentication
INFO  app=edgex-security-proxy-setup source=secret.go:68 msg="Attempting to create secret client"
INFO  app=edgex-security-proxy-setup source=secret.go:73 msg="Created SecretClient"
INFO app=edgex-security-proxy-setup source=config.go:494 msg="Using local configuration from file (3 envVars overri
INFO app=edgex-security-proxy-setup source=secrets.go:322 msg="kick off token renewal with interval: 30m0s"
INFO  app=edgex-security-proxy-setup source=service.go:103 msg="the service on http://127.0.0.1:8100/status is up su
ERROR  app=edgex-security-proxy-setup source=service.go:361 msg="proxy service for scheduler returned status 401"
ERROR app=edgex-security-proxy-setup source=init.go:53 msg="proxy service for scheduler returned status 401"
snap.edgexfoundry.security-proxy-setup.service: Main process exited, code=exited, status=1/FAILURE
snap.edgexfoundry.security-proxy-setup.service: Failed with result 'exit-code'.
Failed to start Service for snap application edgexfoundry.security-proxy-setup.
```

File: snap/local/runtime-helpers/bin/kong-daemon.sh
Comment (jim-wang-intel) [reply snap/local/runtime-helpers/bin/kong-daemon.sh:15]: it should, but i think i can add `|| true` at the end of this command just in case then

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:128]: not sure about the order, i just copied whatever has done in the compose file to the snap here.  @beaufrusetta maybe you can help to answer this better?

File: snap/local/runtime-helpers/bin/kong-daemon.sh
Comment (beaufrusetta) [reply snap/local/runtime-helpers/bin/kong-daemon.sh:15]: The "kong config db_import <file>" command does not return a clean exit code even though it successfully imports. It returns like "6" or something like that - the LUA code behind the scenes is where I found that. 

To get around that, so that the script will continue, I added the `|| true` to the end of the command, and it carries through gracefully. If the import fails, it will continue as if it were good...however, if the import does fail, there are going to be failures immediately down the line when proxy-setup attempts to start - so it doesn't concern me that much because if it does a fail/continue/fail because it won't end up in an exposed middle state - it will fail to setup the Admin API completely and proxy-setup won't be able to complete successfully.

File: snap/snapcraft.yaml
Comment (beaufrusetta) [reply snap/snapcraft.yaml:128]: Port 8100 doesn't need to be exposed externally (that is my fault) - just needs to be available to the local container/snap network in order for proxy-setup to verify that it is up and running so that it can continue setting up services/routes (proxy-setup was previously pinging the admin port to see if it was up, when it should have been using the status port all along). I have not tested binding it to the local Docker network, however, it **_should_** work. We should probably open up a different issue and I can quickly modify/fix that.

The DNS order was modified because of an issue with a Windows based host DNS issue. Keeping the default order would cause the first API request to the admin API to take min of 20 secs to complete, and then others would complete as expected. Over a certain amount of time, that would then reset, and the next API request coming in would exhibit the same behavior. Changing the order did not impact non-Windows based systems, so implementing it across the board seemed pretty low risk.

There are documented cases of this happening in other projects that utilize Kong containers on Windows hosts, and I verified that using my various machines. The work-around that was found in one of the documented issues was to remove the SRV record check by manipulating the Kong DNS checking order via environment variable. Once we did that, the issue went away.

(Thanks to @rsdmike for helping to expose and find a solution for that DNS issue)

Comment (beaufrusetta) [reply snap/snapcraft.yaml:128]: Actually - in the latest version of the `edgex-compose` code, I bound 8100 to the local network through the `ports` directive in the compose file. Setting `KONG_STATUS_LISTEN` is still set to 0.0.0.0:8100, however, Docker compose cuts off the ports access when it creates the container.

Comment (jim-wang-intel) [reply snap/snapcraft.yaml:128]: thanks, @beaufrusetta . I"ll  try to test 8100 port on snap locally here to see if works.

Comment (tonyespy) [reply snap/snapcraft.yaml:128]: I just updated KONG_STATUS_LISTEN to use localhost in the snap and everything comes up just fine.

That said, I'm still seeing the 401 errors after a reboot. The failing service always seems to be different, sometimes its core-data, sometimes -metadata, -scheduler, ... The only consistent is it always seems to be the first that fails. 

I tried adding a 120s sleep before security-proxy-setup starts to rule out a race condition between security-secretstore-setup, and I still get the same failures.

Comment (jim-wang-intel) [reply snap/snapcraft.yaml:128]: > I just updated `KONG_STATUS_LISTEN` to use localhost in the snap and everything comes up just fine.
> 
> That said, I'm still seeing the 401 errors after a reboot. The failing service always seems to be different, sometimes its core-data, sometimes -metadata, -scheduler, ... The only consistent is it always seems to be the first that fails.
> 
> I tried adding a 120s sleep before security-proxy-setup starts to rule out a race condition between security-secretstore-setup, and I still get the same failures.

yea, I also observed the same thing- it is really weird issue in snap. However, when I tried `snap stop` services first right after `snap install`, the 401 error seems gone away. So this is kind of unique in snap as I didn't see anything like this in docker-compose environment.   @tonyespy  any thoughts?

Comment (hutchic) [reply snap/snapcraft.yaml:128]: security wg meeting discussion: we believe this is caused by running [initKongService](https://github.com/edgexfoundry/edgex-taf/issues/403#issuecomment-839338389) on an already initialized Kong

As a way to validate this is the error we could try querying the Kong admin api or run a kong config db_export maybe.

Are we erroniously starting proxy setup?

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml]: changed to bind to localhost only

File: snap/local/runtime-helpers/bin/kong-daemon.sh
Comment (jim-wang-intel) [reply snap/local/runtime-helpers/bin/kong-daemon.sh]: added `|| true`

File: snap/snapcraft.yaml
Comment (jim-wang-intel) [reply snap/snapcraft.yaml:136]: added comments about `KONG_DNS_ORDER` based on @beaufrusetta 's feedback- thanks to @beaufrusetta !

**Review by tonyespy — DISMISSED** [review]
Based upon our discussion at this week's security WG meeting, I'm approving this PR.

We'll continue investigation of the 401 on restart as a part of the effort to ensure robustness of security services across restarts (@lenny-intel do we have an issue to track this work for Ireland).

**Note** - the branch is out-of-sync with master, so needs a rebase before we merge.

**Review by beaufrusetta — APPROVED** [review]
LGTM

**Comment by jim-wang-intel** [comment]
recheck

**Review by hutchic — APPROVED** [review]
looks sensible

**Comment by jim-wang-intel** [comment]
> looks sensible

Thanks!

**Pull request merged**
